hermes/bin/src/packaging/package/mod.rs

```diff
@@ -0,0 +1,358 @@
+            Ok(())
+        } else {
+            Err(anyhow::anyhow!("Directory '{name}' not found"))
+        }
+    }
+
+    /// Get package file reader if present.
+    /// Return error if not possible get a byte reader.
+    pub(crate) fn get_file_reader(&self, name: &str) -> anyhow::Result<Option<impl Read>> {
+        get_package_file_reader(name, &self.0)
+    }
+
+    /// Calculates file hash, if present
+    pub(crate) fn get_file_hash(&self, name: &str) -> anyhow::Result<Option<Blake2b256>> {
+        let mut hasher = Blake2b256Hasher::new();
+
+        Ok(calculate_package_file_hash(name, &self.0, &mut hasher)?.then(|| hasher.finalize()))
+    }
+
+    /// Calculates recursively package directory contents hash including file contents and
+    /// file names.
+    pub(crate) fn get_dir_hash(&self, name: &str) -> anyhow::Result<Option<Blake2b256>> {
+        let mut hasher = Blake2b256Hasher::new();
+        Ok(calculate_package_dir_hash(name, &self.0, &mut hasher)?.then(|| hasher.finalize()))
+    }
+}
+
+/// Copy resource to hdf5 package.
+fn copy_file_to_package(
+    resource: &impl ResourceTrait, name: &str, package: &hdf5::Group,
+) -> anyhow::Result<()> {
+    let mut reader = resource.get_reader()?;
+    let mut resource_data = Vec::new();
+    reader.read_to_end(&mut resource_data)?;
+    if resource_data.is_empty() {
+        return Err(anyhow::anyhow!("Resource {} is empty", resource.name()?));
+    }
+
+    let ds_builder = package.new_dataset_builder();
+    enable_compression(ds_builder)
+        .with_data(&resource_data)
+        .create(name)?;
+
+    Ok(())
+}
+
+/// Copy resource dir to hdf5 package recursively.
+fn copy_dir_recursively_to_package(
+    resource: &impl ResourceTrait, name: &str, package: &hdf5::Group,
+) -> anyhow::Result<()> {
+    let package = package.create_group(name)?;
+
+    let mut errors = Errors::new();
+    for resource in resource.get_directory_content()? {
+        if resource.is_dir() {
+            copy_dir_recursively_to_package(&resource, &resource.name()?, &package).unwrap_or_else(
+                |err| {
+                    match err.downcast::<Errors>() {
+                        Ok(errs) => errors.merge(errs),
+                        Err(err) => errors.add_err(err),
+                    }
+                },
+            );
+        }
+        if resource.is_file() {
+            copy_file_to_package(&resource, &resource.name()?, &package)
+                .unwrap_or_else(|err| errors.add_err(err));
+        }
+    }
+    errors.return_result(())
+}
+
+/// Get package file reader if present.
+/// Return error if not possible get a byte reader.
+fn get_package_file_reader(name: &str, package: &hdf5::Group) -> anyhow::Result<Option<impl Read>> {
+    let reader = package
+        .dataset(name)
+        .ok()
+        .map(|ds| ds.as_byte_reader())
+        .transpose()?;
+    Ok(reader)
+}
+
+/// Buffer size for hash calculation.
+/// 1024 * 1024 = 1MB.
```

Comment:
Can we exceed the size of the buffer? And if so, what will happen?

Reply:
No, this constant is used to setup buffer vector, which is used internally inside `calculate_package_file_hash` with the `Reader::read` function.